src/odemis/driver/test/lakeshore_test.py

```diff
@@ -0,0 +1,124 @@
+
+import logging
+from odemis.driver import lakeshore
+import odemis.model as model
+from odemis.util import test
+import os
+import time
+import unittest
+
+logging.getLogger().setLevel(logging.DEBUG)
+logging.basicConfig(format="%(asctime)s  %(levelname)-7s %(module)s:%(lineno)d %(message)s")
+
+# Export TEST_NOHW=1 to force using only the simulator and skipping test cases
+# needing real hardware
+TEST_NOHW = (os.environ.get("TEST_NOHW", 0) != 0)  # Default to Hw testing
+
+# arguments used for the creation of basic components
+CONFIG = {"name": "Lakeshore Test",
+          "role": "temperature-controller",
+          "port": "/dev/ttyUSB0",
+          "sensor_input": 'b',
+          "output_channel": 2,
+}
+
+# arguments used for the creation of basic components
+if TEST_NOHW:
+    CONFIG["port"] = "/dev/fake"
+
+
+class TestLakeshore(unittest.TestCase):
+    """
+    Tests cases for the Lakeshore 335 Temperature PID
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        cls.dev = lakeshore.Lakeshore(**CONFIG)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.dev.terminate()  # free up socket.
+
+    def test_simple(self):
+        """
+        Test some simple functions
+        """
+        return
+        logging.debug("Test target temperature setting")
+
+        self.assertNotEqual(self.dev.temperature.value, 0)
+        old_val = self.dev.targetTemperature.value
+
+        temps = (-94.20, -90, 0)  # in C
+        for temp in temps:
+            self.dev.targetTemperature.value = temp
+            self.assertAlmostEqual(self.dev.targetTemperature.value, temp)
+            time.sleep(1.0)
+
+        self.dev.targetTemperature.value = old_val
+
+        logging.debug("Test heating range")
+
+        # set heating range
+        old_val = self.dev.heating.value
+        self.dev.heating.value = 0
+        self.assertEqual(self.dev.heating.value, 0)
+        self.dev.heating.value = 1
+        self.assertEqual(self.dev.heating.value, 1)
+        self.dev.heating.value = 2
+        self.assertEqual(self.dev.heating.value, 2)
+        with self.assertRaises(IndexError):
+            self.dev.heating.value = 5
+        self.dev.heating.value = old_val
+
+    def test_temperature_poll(self):
+        return
+        # Test that the temperature polls.
+        # Designed to work with the simulator - the temperature will not be exactly the same
+        logging.debug("Test stasis temperature polling")
+        temp1 = self.dev.temperature.value
+        time.sleep(3)
+        temp2 = self.dev.temperature.value
+        # self.assertNotEqual(temp1, temp2)
+
+    def test_heating_test(self):
+        """
+        Test heating
+        """
+        old = self.dev.targetTemperature.value
+        self.dev.targetTemperature.value = self.dev.temperature.value + 1
+        self.dev.heating.value = 3  # fast heating
+        time.sleep(30)
+        self.assertAlmostEqual(self.dev.temperature.value, self.dev.targetTemperature.value)
```

Comment:
That's not going to work with the simulator, which returns the value +- 5. So either you change the simulator to make it more stable after a while, or increase the temperature change to a larger value (eg +6), and convince the almost equal to have a absolute tolerance of 5°.